File: app/src/main/java/com/hustunique/morii/design/MixActivity.java
package com.hustunique.morii.design;

import static com.hustunique.morii.util.MyApplication.musicTabList;

import android.annotation.SuppressLint;
import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.view.DragEvent;
import android.view.View;
import android.view.animation.AlphaAnimation;
import android.view.animation.Animation;
import android.widget.ImageView;

import androidx.cardview.widget.CardView;
import androidx.constraintlayout.widget.ConstraintLayout;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.bumptech.glide.Glide;
import com.hustunique.morii.database.SoundItemInfo;
import com.hustunique.morii.edit.EditActivity;
import com.hustunique.morii.home.MusicDiaryItem;
import com.hustunique.morii.util.AudioExoPlayerUtil;
import com.hustunique.morii.util.BaseActivity;
import com.hustunique.morii.util.MyApplication;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import morii.R;


public class MixActivity extends BaseActivity {
    @SuppressLint("StaticFieldLeak")
    private static int _last_drag_position;
    private RecyclerView recyclerView;
    private androidx.cardview.widget.CardView backLayout, completeLayout;
    private static ConstraintLayout delete_area;
    private ImageView helpMessage;
    private ImageView cardImage;
    private static final String TAG = "MixActivity";
    private ImageView helpButton;
    private static int playbackStatus = -1;
    private final List<ConstraintLayout> squareLayoutList = new ArrayList<>(9);
    private final List<ImageView> squareList = new ArrayList<>(9);
    private static Animation appear = new AlphaAnimation(0f, 1f);
    private static Animation fade = new AlphaAnimation(1f, 0f);
    private final Map<ConstraintLayout, SoundItem> constraintLayoutSoundItemMap = new HashMap<>(9);
    private final Map<Integer, Integer> positionSoundItemIdMap = new HashMap<>(9);
    private ImageView playbackIcon;
    private CardView playbackButton;
    private final int animationDuration = 150;
    private final List<View> hints = new ArrayList<>();

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_mix);
        initUI();
    }

    private void initUI() {
        viewBinding();
        MusicDiaryItem diary = (MusicDiaryItem) getIntent().getSerializableExtra("diary");
        Glide.with(this).load(musicTabList.get(diary.getMusicTabId()).getImageResId()).into(cardImage);
        helpButton.setOnClickListener(v -> {
            if (helpMessage.getVisibility() == View.GONE)
                startFadeIn(helpMessage);
            else startFadeOut(helpMessage);
        });
        helpMessage.setOnClickListener(v -> {
            if (v.getVisibility() == View.VISIBLE)
                startFadeOut(helpMessage);
        });
        playbackButton.setOnClickListener(v -> {
            if (playbackStatus == 1) {
                AudioExoPlayerUtil.startAllPlayers();
                Glide.with(this).load(R.drawable.outline_pause_24).into(playbackIcon);
            } else {
                AudioExoPlayerUtil.pauseAllPlayers();
                Glide.with(this).load(R.drawable.round_play_arrow_24).into(playbackIcon);
            }
            playbackStatus = -playbackStatus;
        });
        completeLayout.setOnClickListener(v -> {
            Intent intent = new Intent(this, EditActivity.class);
            diary.getSoundItemInfoList().clear();
            for (Map.Entry<Integer, Integer> entry : positionSoundItemIdMap.entrySet()) {
                diary.getSoundItemInfoList().add(new SoundItemInfo(entry.getKey(), entry.getValue()));
                Log.d("activityData", "key = " + entry.getKey() + " value = " + entry.getValue());
            }
            intent.putExtra("diary", diary);
            startActivity(intent);
        });
        backLayout.setOnClickListener(v -> onBackPressed());
        for (ConstraintLayout constraintLayout : squareLayoutList) {
            constraintLayoutSoundItemMap.put(constraintLayout, null);
        }

        for (ConstraintLayout constraintLayout : constraintLayoutSoundItemMap.keySet()) {
            constraintLayout.setOnDragListener(((v, event) -> {
                int rmposition;
                int position = squareLayoutList.indexOf(constraintLayout);
                int soundItemId;
                int iconID;
                if (event.getAction() == DragEvent.ACTION_DROP) {
                    rmposition = event.getClipData().getItemAt(0).getIntent().getIntExtra("position", -1);
                    if (rmposition != -1) {
                        /**
                         *here to cover a sound item , params may be used here are
                         *  positionSoundItemIdMap.get(rmposition) -- soundItemId
                         *  rmposition ,the position of the soundItem
                         */
                        Log.d(TAG, "3");
                        stopPlayingSoundItem(positionSoundItemIdMap.get(rmposition), rmposition);
                        positionSoundItemIdMap.remove(rmposition);
                        constraintLayoutSoundItemMap.remove(squareLayoutList.get(rmposition));
                        squareLayoutList.get(rmposition).setOnTouchListener(null);
                    }
                    iconID = event.getClipData().getItemAt(0).getIntent().getIntExtra("ImageID",
                            R.drawable.outline_air_24);
                    constraintLayout.setAlpha(1.0f);
                    soundItemId = event.getClipData().getItemAt(0).getIntent().getIntExtra("indexOfSoundItem", 0);
                    squareList.get(position).setImageResource(iconID);
                    positionSoundItemIdMap.put(position, soundItemId);
                    constraintLayoutSoundItemMap.put(constraintLayout, MyApplication.soundItemList.get(soundItemId));
                    constraintLayout.setOnTouchListener(new StartDrag(position, true, soundItemId));
                    /**
                     *here to start a sound item , params may be used here are
                     * soundItemId  -- soundItemId
                     *  position ,the position of the soundItem
                     */
                    Log.d(TAG, "2");
                    startPlayingSoundItem(soundItemId, position);
                    return true;
                }
                return event.getAction() == DragEvent.ACTION_DRAG_STARTED;
            }));
        }
        /*cardImage.setOnDragListener(new View.OnDragListener() {
            @Override
            public boolean onDrag(View v, DragEvent event) {
                switch (event.getAction()){
                    case DragEvent.ACTION_DRAG_STARTED:
                        Log.d("debug_drag","start");
                        _last_drag_position = (Integer) event.getLocalState();
                        return true;
                        case DragEvent.ACTION_DRAG_ENDED:
                            Log.d("debug_drag","end "+_last_drag_position);
                            if(!event.getResult()&&_last_drag_position!=-1){
                                squareLayoutList.get(_last_drag_position).setAlpha(1.0f);
                                hideDeleteArea();
                            }
                            return true;
                    default:
                        return true;
                }
            }
        });

         */
        delete_area.setOnDragListener((v, event) -> {
            if (event.getAction() == DragEvent.ACTION_DROP) {
                int rmposition = event.getClipData().getItemAt(0).getIntent().getIntExtra("position", -1);
                if (rmposition != -1) {
                    /**
                     *here to stop a sound item , params may be used here are
                     *  positionSoundItemIdMap.get(rmposition) -- soundItemId
                     *  rmposition ,the position of the soundItem
                     */
                    Log.d(TAG, "1");
                    stopPlayingSoundItem(positionSoundItemIdMap.get(rmposition), rmposition);
                    positionSoundItemIdMap.remove(rmposition);
                    Glide.with(this).load(R.drawable.square_transparent).into(squareList.get(rmposition));
                    //squareList.get(rmposition).setImageResource(R.drawable.square_transparent);
                    constraintLayoutSoundItemMap.remove(squareLayoutList.get(rmposition));
                    squareLayoutList.get(rmposition).setOnTouchListener(null);
                    StartDrag.makeVibrate();
                    return true;
                }
                return false;
            } else if (event.getAction() == DragEvent.ACTION_DRAG_ENDED) {
                if (!event.getResult() && _last_drag_position != -1) {
                    squareLayoutList.get(_last_drag_position).setAlpha(1.0f);
                }
                StartDrag.makeVibrate();
                v.setAlpha(0f);
                delete_area.startAnimation(fade);
                onItemDropped();
            } else if (event.getAction() == DragEvent.ACTION_DRAG_STARTED) {
                _last_drag_position = (Integer) event.getLocalState();
                Log.d("debug_drag", "start" + _last_drag_position);
                if (_last_drag_position != -1) {
                    delete_area.startAnimation(appear);
                    v.setAlpha(1f);
                }
                onItemDragged();
            }
            return true;
        });
        setRecyclerView();
    }

    private void viewBinding() {
        recyclerView = findViewById(R.id.recyclerView03);
        cardImage = findViewById(R.id.imageView_card);
        delete_area = findViewById(R.id.delete_area);
        backLayout = findViewById(R.id.backLayout_select_mix);
        completeLayout = findViewById(R.id.okay_mix);
        squareList.add(findViewById(R.id.square_image1));
        squareList.add(findViewById(R.id.square_image2));
        squareList.add(findViewById(R.id.square_image3));
        squareList.add(findViewById(R.id.square_image4));
        squareList.add(findViewById(R.id.square_image5));
        squareList.add(findViewById(R.id.square_image6));
        squareList.add(findViewById(R.id.square_image7));
        squareList.add(findViewById(R.id.square_image8));
        squareList.add(findViewById(R.id.square_image9));
        squareLayoutList.add(findViewById(R.id.square1));
        squareLayoutList.add(findViewById(R.id.square2));
        squareLayoutList.add(findViewById(R.id.square3));
        squareLayoutList.add(findViewById(R.id.square4));
        squareLayoutList.add(findViewById(R.id.square5));
        squareLayoutList.add(findViewById(R.id.square6));
        squareLayoutList.add(findViewById(R.id.square7));
        squareLayoutList.add(findViewById(R.id.square8));
        squareLayoutList.add(findViewById(R.id.square9));
        hints.add(findViewById(R.id.hintText1));
        hints.add(findViewById(R.id.hintText2));
        hints.add(findViewById(R.id.hintText3));
        hints.add(findViewById(R.id.hintText4));
        hints.add(findViewById(R.id.hintBG));
        for (View v : hints) {
            v.setVisibility(View.INVISIBLE);
        }
        playbackButton = findViewById(R.id.playbackButton);
        playbackIcon = findViewById(R.id.playbackButton_image);
        appear.setDuration(animationDuration);
        appear.setFillAfter(true);
        fade.setDuration(animationDuration);
        fade.setFillAfter(true);
        fade.setAnimationListener(new Animation.AnimationListener() {
            @Override
            public void onAnimationStart(Animation animation) {

            }

            @Override
            public void onAnimationEnd(Animation animation) {
                Log.d(TAG, "onAnimationEnd: ");
            }

            @Override
            public void onAnimationRepeat(Animation animation) {

            }
        });
        helpButton = findViewById(R.id.helpButton);
        helpMessage = findViewById(R.id.helpMessage);
    }

    private void setRecyclerView() {
        WhiteNoiseAdapter adapter = new WhiteNoiseAdapter(this);
        LinearLayoutManager manager = new LinearLayoutManager(this, LinearLayoutManager.HORIZONTAL, false);
        recyclerView.setLayoutManager(manager);
        recyclerView.setAdapter(adapter);
        recyclerView.setOnDragListener(new View.OnDragListener() {
            @Override
            public boolean onDrag(View v, DragEvent event) {
                return false;
            }
        });
    }


    private void stopPlayingSoundItem(int soundItemId, int position) {
        AudioExoPlayerUtil.stopPlayingSoundItem(position);
    }

    private void startPlayingSoundItem(int soundItemId, int position) {
        AudioExoPlayerUtil.setSoundPlayer(soundItemId, position);
        if (playbackStatus == -1)
            AudioExoPlayerUtil.startSoundPlayer(position);
    }

    @Override
    protected void onResume() {
        super.onResume();
        if (!AudioExoPlayerUtil.isPlaying()) {
            playbackStatus = 1;
            Glide.with(this).load(R.drawable.round_play_arrow_24).into(playbackIcon);
        } else {
            playbackStatus = -1;
            Glide.with(this).load(R.drawable.outline_pause_24).into(playbackIcon);
        }
    }

    @Override
    public void onBackPressed() {
        super.onBackPressed();
        AudioExoPlayerUtil.stopAllSoundPlayers();
        AudioExoPlayerUtil.resetAllSoundPlayers();
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();

    }

    private void onItemDragged() {
        for (View v : hints) {
            startFadeIn(v);
            appear = new AlphaAnimation(0f, 1f);
            appear.setDuration(animationDuration);
        }
    }

    private void onItemDropped() {
        for (View v : hints) {
            if (v.getVisibility() == View.VISIBLE) {
                startFadeOut(v);
                fade = new AlphaAnimation(1f, 0f);
                fade.setDuration(150);
            }
        }
    }

    private void startFadeOut(View v) {
        if (v.getVisibility() == View.VISIBLE) {

            v.startAnimation(fade);
            v.setVisibility(View.GONE);
        }
    }

    private void startFadeIn(View v) {
        v.setVisibility(View.VISIBLE);
        v.startAnimation(appear);
    }
}

File: app/src/main/java/com/hustunique/morii/design/MixActivity.kt
package com.hustunique.morii.design

import android.content.Intent
import android.annotation.SuppressLint
import com.hustunique.morii.util.BaseActivity
import androidx.constraintlayout.widget.ConstraintLayout
import com.hustunique.morii.util.AudioExoPlayerUtil
import android.os.Bundle
import morii.R
import com.bumptech.glide.Glide
import com.hustunique.morii.home.MusicDiaryItem
import com.hustunique.morii.util.MyApplication
import com.hustunique.morii.edit.EditActivity
import com.hustunique.morii.database.SoundItemInfo
import android.view.animation.AlphaAnimation
import android.util.Log
import android.view.*
import android.view.View.OnDragListener
import android.view.animation.Animation
import android.widget.*
import androidx.recyclerview.widget.LinearLayoutManager
import morii.databinding.ActivityMixBinding
import java.util.ArrayList
import java.util.HashMap

class MixActivity : BaseActivity() {
    private val squareLayoutList: MutableList<ConstraintLayout> = ArrayList(9)
    private val squareList: MutableList<ImageView> = ArrayList(9)
    private val constraintLayoutSoundItemMap: MutableMap<ConstraintLayout, SoundItem?> = HashMap(9)
    private val positionSoundItemIdMap: MutableMap<Int, Int> = HashMap(9)
    private val animationDuration = 150
    private val hints: MutableList<View> = ArrayList()
    private lateinit var binding: ActivityMixBinding

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityMixBinding.inflate(layoutInflater)
        setContentView(binding.root)
        initUI()
    }

    private fun initUI() {
        initLists()
        val diary = intent.getSerializableExtra("diary") as MusicDiaryItem
        Glide.with(this)
            .load(MyApplication.musicTabList[diary.musicTabId].imageResId)
            .into(binding.imageViewCard)
        binding.helpButton.setOnClickListener { v: View ->
            if (binding.helpMessage.visibility == View.GONE) startFadeIn(
                binding.helpMessage
            ) else startFadeOut(binding.helpMessage)
        }
        binding.helpMessage.setOnClickListener { v: View ->
            if (v.visibility == View.VISIBLE) startFadeOut(
                binding.helpMessage
            )
        }
        binding.playbackButton.setOnClickListener { v: View ->
            if (playbackStatus == 1) {
                AudioExoPlayerUtil.startAllPlayers()
                Glide.with(this).load(R.drawable.outline_pause_24).into(
                    binding.playbackButtonImage
                )
            } else {
                AudioExoPlayerUtil.pauseAllPlayers()
                Glide.with(this).load(R.drawable.round_play_arrow_24).into(
                    binding.playbackButtonImage
                )
            }
            playbackStatus = -playbackStatus
        }
        binding.okayMix.setOnClickListener { v: View ->
            val intent = Intent(this, EditActivity::class.java)
            diary.soundItemInfoList.clear()
            for ((key, value) in positionSoundItemIdMap) {
                diary.soundItemInfoList.add(SoundItemInfo(key, value))
                Log.d("activityData", "key = $key value = $value")
            }
            intent.putExtra("diary", diary)
            startActivity(intent)
        }
        binding.backLayoutSelectMix.setOnClickListener { v: View -> onBackPressed() }
        for (constraintLayout in squareLayoutList) {
            constraintLayoutSoundItemMap[constraintLayout] = null
        }
        for (constraintLayout in constraintLayoutSoundItemMap.keys) {
            constraintLayout.setOnDragListener(label@ OnDragListener { v: View, event: DragEvent ->
                val rmposition: Int
                val position = squareLayoutList.indexOf(constraintLayout)
                val soundItemId: Int
                val iconID: Int
                if (event.action == DragEvent.ACTION_DROP) {
                    rmposition = event.clipData.getItemAt(0).intent.getIntExtra("position", -1)
                    if (rmposition != -1) {
                        Log.d(TAG, "3")
                        stopPlayingSoundItem(positionSoundItemIdMap[rmposition]!!, rmposition)
                        positionSoundItemIdMap.remove(rmposition)
                        constraintLayoutSoundItemMap.remove(squareLayoutList[rmposition])
                        squareLayoutList[rmposition].setOnTouchListener(null)
                    }
                    iconID = event.clipData.getItemAt(0).intent.getIntExtra(
                        "ImageID",
                        R.drawable.outline_air_24
                    )
                    constraintLayout.alpha = 1.0f
                    soundItemId =
                        event.clipData.getItemAt(0).intent.getIntExtra("indexOfSoundItem", 0)
                    squareList[position].setImageResource(iconID)
                    positionSoundItemIdMap[position] = soundItemId
                    constraintLayoutSoundItemMap[constraintLayout] =
                        MyApplication.soundItemList[soundItemId]
                    constraintLayout.setOnTouchListener(StartDrag(position, true, soundItemId))
                    Log.d(TAG, "2")
                    startPlayingSoundItem(soundItemId, position)
                    return@OnDragListener true
                }
                event.action == DragEvent.ACTION_DRAG_STARTED
            })
        }
        binding.deleteArea.setOnDragListener { v: View, event: DragEvent ->
            if (event.action == DragEvent.ACTION_DROP) {
                val rmposition = event.clipData.getItemAt(0).intent.getIntExtra("position", -1)
                if (rmposition != -1) {
                    Log.d(TAG, "1")
                    stopPlayingSoundItem(positionSoundItemIdMap[rmposition]!!, rmposition)
                    positionSoundItemIdMap.remove(rmposition)
                    Glide.with(this).load(R.drawable.square_transparent)
                        .into(squareList[rmposition])
                    constraintLayoutSoundItemMap.remove(squareLayoutList[rmposition])
                    squareLayoutList[rmposition].setOnTouchListener(null)
                    StartDrag.makeVibrate()
                    return@setOnDragListener true
                }
                return@setOnDragListener false
            } else if (event.action == DragEvent.ACTION_DRAG_ENDED) {
                if (!event.result && previousPosition != -1) {
                    squareLayoutList[previousPosition].alpha = 1.0f
                }
                StartDrag.makeVibrate()
                v.alpha = 0f
                binding.deleteArea.startAnimation(fade)
                onItemDropped()
            } else if (event.action == DragEvent.ACTION_DRAG_STARTED) {
                previousPosition = event.localState as Int
                Log.d("debug_drag", "start$previousPosition")
                if (previousPosition != -1) {
                    binding.deleteArea.startAnimation(appear)
                    v.alpha = 1f
                }
                onItemDragged()
            }
            true
        }
        setRecyclerView()
    }

    private fun initLists() {
        squareList.add(binding.squareImage1)
        squareList.add(binding.squareImage2)
        squareList.add(binding.squareImage3)
        squareList.add(binding.squareImage4)
        squareList.add(binding.squareImage5)
        squareList.add(binding.squareImage6)
        squareList.add(binding.squareImage7)
        squareList.add(binding.squareImage8)
        squareList.add(binding.squareImage9)
        squareLayoutList.add(binding.square1)
        squareLayoutList.add(binding.square2)
        squareLayoutList.add(binding.square3)
        squareLayoutList.add(binding.square4)
        squareLayoutList.add(binding.square5)
        squareLayoutList.add(binding.square6)
        squareLayoutList.add(binding.square7)
        squareLayoutList.add(binding.square8)
        squareLayoutList.add(binding.square9)
        hints.add(binding.hintText1)
        hints.add(binding.hintText2)
        hints.add(binding.hintText3)
        hints.add(binding.hintText4)
        hints.add(binding.hintBG)
        for (v in hints) {
            v.visibility = View.INVISIBLE
        }
        appear.duration = animationDuration.toLong()
        appear.fillAfter = true
        fade.duration = animationDuration.toLong()
        fade.fillAfter = true
        fade.setAnimationListener(object : Animation.AnimationListener {
            override fun onAnimationStart(animation: Animation) {}
            override fun onAnimationEnd(animation: Animation) {
                Log.d(TAG, "onAnimationEnd: ")
            }

            override fun onAnimationRepeat(animation: Animation) {}
        })

    }

    private fun setRecyclerView() {
        val adapter = WhiteNoiseAdapter(this)
        val manager = LinearLayoutManager(this, LinearLayoutManager.HORIZONTAL, false)
        binding.soundItemRecyclerView.apply {
            layoutManager = manager
            setAdapter(adapter)
            setOnDragListener { _, _ -> false }
        }
    }

    private fun stopPlayingSoundItem(soundItemId: Int, position: Int) {
        AudioExoPlayerUtil.stopPlayingSoundItem(position)
    }

    private fun startPlayingSoundItem(soundItemId: Int, position: Int) {
        AudioExoPlayerUtil.setSoundPlayer(soundItemId, position)
        if (playbackStatus == -1) AudioExoPlayerUtil.startSoundPlayer(position)
    }

    override fun onResume() {
        super.onResume()
        if (!AudioExoPlayerUtil.isPlaying) {
            playbackStatus = 1
            Glide.with(this).load(R.drawable.round_play_arrow_24).into(
                binding.playbackButtonImage
            )
        } else {
            playbackStatus = -1
            Glide.with(this).load(R.drawable.outline_pause_24).into(
                binding.playbackButtonImage
            )
        }
    }

    override fun onBackPressed() {
        super.onBackPressed()
        AudioExoPlayerUtil.stopAllSoundPlayers()
        AudioExoPlayerUtil.resetAllSoundPlayers()
    }

    private fun onItemDragged() {
        for (v in hints) {
            startFadeIn(v)
            appear = AlphaAnimation(0f, 1f)
            appear.duration = animationDuration.toLong()
        }
    }

    private fun onItemDropped() {
        for (v in hints) {
            if (v.visibility == View.VISIBLE) {
                startFadeOut(v)
                fade = AlphaAnimation(1f, 0f)
                fade.duration = 150
            }
        }
    }

    private fun startFadeOut(v: View) {
        if (v.visibility == View.VISIBLE) {
            v.startAnimation(fade)
            v.visibility = View.GONE
        }
    }

    private fun startFadeIn(v: View) {
        v.visibility = View.VISIBLE
        v.startAnimation(appear)
    }

    companion object {
        @SuppressLint("StaticFieldLeak")
        private var previousPosition = 0
        private const val TAG = "MixActivity"
        private var playbackStatus = -1
        private var appear: Animation = AlphaAnimation(0f, 1f)
        private var fade: Animation = AlphaAnimation(1f, 0f)
    }
}